### ISSUE TITLE
Jenkins PR builder Plan

### ISSUE THREAD

**Issue by askldjd** [department-of-veterans-affairs/caseflow#2573]
## Things to do / think about:

- Inject `JENKINS_PR_BUILDER` env to Rails during PR build. This way, we can avoid setting `--no-sandbox` in `karma.conf.js` and `rails_helper.rb` in developer's machine.

- For the 2 tests that are skipped, how do we fix it?

- The `docker-compose.yml` file in Caseflow still doesn't quite work for me. What should the developer experience be? 

- We need a pipeline to build the base docker image periodically and push it to the repository. We may consider versioning the images too.

- Can we confirm that the Jenkins pipeline can target specific filename such as `Jenksinfile.pr`?

- Should we coordinate with @joofsh and @shanear to optimize the build speed further? (e.g. parallelizing spec)

## Roll out thoughts:

- Merge to master, and create a multi-branch pipeline to build all Caseflow PRs.
- Verify that it is stable, and tie in Jenkins notification to GH
- Deprecate Travis.

**Comment by mogrenAtWork** [comment]
I agree with the rollout plan, addressing the 'things to think about' section in order below:
1. I'm not sure how injecting the `JENKINS_PR_BUILDER` env would fix this problem. I'd love to fix this problem though as the `no-sandbox` option is not ideal. Could you explain this solution a little more?
1. We absolutely need to address these. I can't put my finger on the exact problem, from speaking with Alan last night it's probably a timing issue where Selenium isn't properly waiting for an element to load on the page or the Javascript test configuration isn't perfect. We should create separate tickets for these to debug.
1. The `docker-compose` environment should contain all dependencies for devs to get started quickly. The steps to initialize the caseflow project should be:
    1. Install Docker and checkout the project
    1. Run `docker-composer` to create a environment with all initial dependencies and the project source should be mapped to the local file system to pick up any changes
    1. Run `docker-compose run web rake db:setup` to initialize the database
    1. `docker-compose` should be running all services from the start and you should be able to access any of these (redis, postgres, web app) from your local computer
         - Devs should be able to debug using byebug from within the `docker-compose` environment, use psql and redis-cli from their local system to connect to the postgres service to run queries 
1. Versioning the images would be awesome. If we could do this, it would also eliminate the need for our local Gem and NPM repos that we have running.
1. I think we'll know the answer to this question once #2448 is merged to master. It specifies the new Jenkinsfile as `Jenkinsfile.new` so it doesn't mess with the existing pipeline. We should be able to create a multi-branch job in the new Jenkins instance to use this filename. I had problems initially because the multi-branch job looks for the filename in `master` instead of the branches that it's executing the job against.
1. Absolutely. There also might be a way we can add timestamps to the console or test executions to find the largest bottlenecks and biggest wins

**Comment by askldjd** [comment]
> I'm not sure how injecting the JENKINS_PR_BUILDER env would fix this problem. I'd love to fix this problem though as the no-sandbox option is not ideal. Could you explain this solution a little more?

I wasn't thinking of anything fancy. By defining the env, we can do some `if/else` magic to skip the `no-sandbox` option if the env variable is defined.

Regarding docker-compose setup, are you expecting dev to mount their folder to the local folder? I am still a bit confused about how I would set it up.

**Comment by tejans24** [comment]
@mogrenAtWork For 3, is it meant for actively developing on docker containers as part of the development flow, or is it mostly for debuting tests and other setup related things?

**Comment by mogrenAtWork** [comment]
@tejans24 all of it. I think debugging tests and setup is included in development flow. Also, this isn't mandatory, it's optional for quick setup.

**Comment by kierachell** [comment]
I will do the two issues that do not block this from working:
- New issue for `docker-compose` for developers
- New issue for `docker image bake` on current Jenkins

Optimizations to add:
- Parallel test
- Yarn for faster build?
- Fail-fast for failing parallel stages

**Comment by tejans24** [comment]
@mogrenAtWork gotcha, thanks. I think it would be nice to actively use docker as part of the development flow. It think it might be helpful for the db container to have a persistent volume attached and hot reload for react to work. But that could be part of a future iteration as well.

**Comment by mogrenAtWork** [comment]
@tejans24 good points. As of now, the DB container mounts from a `pgdata` directory on your local file system within the project, so this will be persisted. Hot-reload should work as well. I'll clarify those in the description.

**Comment by mogrenAtWork** [comment]
@askldjd 

> Regarding docker-compose setup, are you expecting dev to mount their folder to the local folder? I am still a bit confused about how I would set it up.

The docker-compose.yml file already has all of the volumes mounted to specific directories within the project, so the dev will not need to alter anything to get this working. These directories are also excluded from git appropriately using `.gitignore`

**Comment by askldjd** [comment]
ok, i will give this a spin later tonight. Thanks

**Comment by mogrenAtWork** [comment]
I've updated the tech spec for the Jenkins migration with the discussion from this thread: https://github.com/department-of-veterans-affairs/appeals-pm/blob/master/caseflow/tech_specs/migrate_to_jenkins_for_all_ci.md

Let me know if anything needs to be changed, otherwise, I'll close this issue out.

**Comment by mogrenAtWork** [comment]
Closing this out
